File: src/js/Classes/Universe.ts
import Body from './Body';

export default class Universe {
  bodies: Body[];
  readonly gravity: number;
  readonly timeStep: number;

  constructor(gravity = .5, timeStep = 8) {
    this.bodies = [];
    this.gravity = gravity;
    this.timeStep = 1 / timeStep;
  }

  checkCollision(b: Body, b2: Body): boolean {
    const d = b.position.sub(b2.position);
    const r = b.radius + b2.radius;
    return d.lengthSq() < r * r;
  }

  computeForces() {
    for (let i = 0; i < this.bodies.length; i += 1) {
      const b1 = this.bodies[i];
      const { acceleration: a1, mass: m1 } = b1;
      a1.set(0);

      for (let j = 0; j < i; j += 1) {
        const b2 = this.bodies[j];
        const { acceleration: a2, mass: m2 } = b2;

        const d = b1.position.sub(b2.position);
        const norm = Math.sqrt(100.0 + d.lengthSq());
        const mag = this.gravity / (norm ** 3);

        a1.set(a1.sub(d.mul(mag * m2)));
        a2.set(a2.add(d.mul(mag * m1)));
      }
    }
  }

  doCollisions() {
    for (let i = 0; i < this.bodies.length; i += 1) {
      const b = this.bodies[i];

      for (let j = 0; j < i; j += 1) {
        const b2 = this.bodies[j];

        if (this.checkCollision(b, b2))
          this.resolveCollision(b, b2);
      }
    }
  }

  doPhysics() {
    const dt = this.timeStep;
    for (const b of this.bodies) {
      b.position = b.position.add(b.velocity.mul(0.5 * dt));
    }
    this.computeForces();

    for (const b of this.bodies) {
      b.velocity = b.velocity.add(b.acceleration.mul(dt));
    }

    for (const b of this.bodies) {
      b.position = b.position.add(b.velocity.mul(0.5 * dt));
    }

    this.doCollisions();
  }

  resolveCollision(b: Body, b2: Body): void {
    const d = b.position.sub(b2.position);
    d.set(d.normalize());

    const v = b2.velocity.sub(b.velocity);
    const dot = d.dot(v);

    if (dot >= 0) {
      const tm = b.mass + b2.mass;
      const c = d.mul(2 * dot / tm);

      b.velocity.set(b.velocity.add(c.mul(b2.mass)));
      b2.velocity.set(b2.velocity.sub(c.mul(b.mass)));
    }
  }
}

File: src/cpp/nbody/universe.cc
#include "universe.h"

namespace nbody {

Universe::Universe(float gravity, float time_index)
  : gravity_(gravity),
  time_index_(time_index == 0 ? throw std::exception("Divide by zero: time_index cannot be zero. (in Universe() constructor)") : time_index)
{}

bool Universe::CheckCollision(const Body& a, const Body& b) {
  Vector delta = a.Subtract(b);
  float radii = a.radius() + b.radius();
  return delta.LengthSquared() < (radii * radii);
}

void Universe::ComputeForces() {
  size_t num_bodies = bodies_.size();
  for (size_t i = 0; i < num_bodies; i++) {
    Body& b1 = *bodies_[i];         // body1
    auto a1 = b1.acceleration();   // accelleration1
    auto m1 = b1.mass();           // mass1
    a1.Set(0);

    for (size_t j = 0; j < i; j++) {
      Body& b2 = *bodies_[j];       // body2
      auto a2 = b2.acceleration(); // accelleration2
      auto m2 = b2.mass();         // mass2

      auto delta = b1.Subtract(b2);
      float norm = sqrt(100.0f + delta.LengthSquared());
      float mag = gravity_ / (norm * norm * norm);

      a1.Set(a1.Subtract(delta.Multiply(mag * m2)));
      a2.Set(a2.Add(delta.Multiply(mag * m1)));
    }
  }
}

void Universe::DoCollisions() {
  size_t num_bodies = bodies_.size();
  for (size_t i = 0; i < num_bodies; i++) {
    Body& b = *bodies_[i];

    for (size_t j = 0; j < i; j++) {
      Body& b2 = *bodies_[j];
      if (CheckCollision(b, b2))
        ResolveCollision(b, b2);
    }
  }
}

// DANGER:
// Body.velocity() (below) is providing a non-const reference -- NOT value.
// i.e., logic below is mutating the bodies' velocity Vector.
// TODO: This is provisional and needs addressing.
void Universe::DoPhysics() {
  const float dt = time_index_;
  for (Body* b : bodies_) {
    (*b).Set((*b).Add((*b).velocity().Multiply(0.5f * dt)));
  }
  ComputeForces();

  for (Body* b : bodies_) {
    (*b).velocity().Set((*b).velocity().Add((*b).acceleration().Multiply(dt)));
  }

  for (Body* b : bodies_) {
    (*b).Set((*b).Add((*b).velocity().Multiply(0.5f * dt)));
  }

  DoCollisions();
}

// DANGER:
// Body.velocity() (below) is providing a non-const reference -- NOT value.
// i.e., logic below is mutating the bodies' velocity Vector.
// TODO: This is provisional and needs addressing.
void Universe::ResolveCollision(Body& a, Body& b) {
  Vector delta = a.Subtract(b);
  delta.Set(delta.Normalize());

  auto v = b.velocity().Subtract(a.velocity());
  auto dot_product = delta.Dot(v);

  if (dot_product >= 0) {
    auto total_mass = a.mass() + b.mass();
    auto c = delta.Multiply(2 * dot_product / total_mass);
    a.velocity().Set(a.velocity().Add(c.Multiply(b.mass())));
    b.velocity().Set(b.velocity().Subtract(c.Multiply(a.mass())));
  }
}

} // namespace nbody
